streaming/base/dataset.py

```diff
@@ -1284,6 +1284,9 @@ def _ready_thread(self, it: _Iterator) -> None:
                 self.prepare_shard(shard_id, False)
             # Wait for a shard file to download completely.
             while self._shard_states[shard_id] != _ShardState.LOCAL:
+                # Background thread or a main process crashed, terminate this thread.
```

Comment:
if i'm understanding correctly, the bug here is that in the case when we're waiting for a shard file to download completely, if there is an exception we won't terminate because we're not checking `self._event.is_set()` which is set on exception. In other cases we do check for this so the threads terminate in other cases. Right?

Reply:
First of all, the `ThreadPoolExecutor` does not provide a utility to stop running tasks, we can update our target task functions to stop running when a thread-safe flag is set and this can be achieved using [**threading.Event**](https://docs.python.org/3/library/threading.html) call. There are two threads that streaming uses, `_prepare_thread()` which downloads the shard file, and `_ready_thread()` which ensures the samples are ready to yield. If samples are not ready to yield, it waits for the shard file to present locally (which `_prepare_thread()` is working on to download the shard file from remote location to a local directory). Now, take a case where the `_prepare_thread()`  crashed due to any reason, in that case, `_prepare_thread()`  sets the threading event and exit. Now, since the `_ready_thread()` is in the infinite loop, there is no way to notify `_ready_thread()` that the `_prepare_thread()` is crashed. One way of doing this is that we check for the `threading.Event` in that infinite while loop and exit as soon as the event is set.